pkg/agent/openflow/multicluster.go

```diff
@@ -0,0 +1,231 @@
+			Action().SetSrcMAC(localGatewayMAC).
+			Action().SetDstMAC(GlobalVirtualMACForMulticluster).
+			Action().SetTunnelDst(tunnelPeer). // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+	)
+	return flows
+}
+
+func (f *featureMulticluster) tunnelClassifierFlow(tunnelOFPort uint32) binding.Flow {
+	return ClassifierTable.ofTable.BuildFlow(priorityHigh).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchInPort(tunnelOFPort).
+		MatchDstMAC(GlobalVirtualMACForMulticluster).
+		Action().LoadRegMark(FromTunnelRegMark).
+		Action().LoadRegMark(RewriteMACRegMark).
+		Action().GotoStage(stageConntrackState).
+		Done()
+}
+
+func (f *featureMulticluster) outputFlow(tunnelOFPort uint32) binding.Flow {
+	return L2ForwardingOutTable.ofTable.BuildFlow(priorityHigh).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchRegMark(RewriteMACRegMark).
+		MatchRegMark(OutputToTunnelRegMark).
+		MatchInPort(tunnelOFPort).
+		Action().OutputInPort().
+		Done()
+}
+
+// snatConntrackFlows generates the flow to match the first packet of a cross-cluster connection.
+func (f *featureMulticluster) snatConntrackFlows(serviceCIDR net.IPNet, localGatewayIP net.IP) []binding.Flow {
+	var flows []binding.Flow
+	ipProtocol := getIPProtocol(localGatewayIP)
+
+	flows = append(flows,
+		// This generates the flow to match the first packet of a Service connection sourced from a local Pod or local
+		// Antrea gateway, and the Endpoint is a Service ClusterIP on remote cluster. Note that, DNAT is performed
+		// on such connection, and SNAT should also be performed on such connection to forward the packets to remote cluster.
+		// This flow is used to mark the connection requiring SNAT by loading ConnSNATCTMark in DNAT CT zone.
+		ServiceMarkTable.ofTable.BuildFlow(priorityHigh).
+			Cookie(f.cookieAllocator.Request(f.category).Raw()).
+			MatchProtocol(ipProtocol).
+			MatchCTStateNew(true).
+			MatchCTStateTrk(true).
+			MatchCTMark(ServiceCTMark).
+			MatchDstIPNet(serviceCIDR).
+			Action().CT(true, ServiceMarkTable.GetNext(), f.dnatCtZones[ipProtocol], nil).
+			LoadToCtMark(ConnSNATCTMark).
+			CTDone().
+			Done(),
+		// This generates the flow to perform SNAT for the cross-cluster Service connections sourced
+		// from local Pods or local Antrea gateway.
+		SNATConntrackCommitTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(f.cookieAllocator.Request(f.category).Raw()).
+			MatchProtocol(ipProtocol).
+			MatchCTStateNew(true).
+			MatchCTStateTrk(true).
+			MatchCTMark(ServiceCTMark).
+			MatchDstIPNet(serviceCIDR).
+			Action().CT(true, SNATConntrackCommitTable.GetNext(), f.snatCtZones[ipProtocol], nil).
+			SNAT(&binding.IPRange{StartIP: localGatewayIP, EndIP: localGatewayIP}, nil).
+			LoadToCtMark(ServiceCTMark).
+			CTDone().
+			Done(),
+		// This generates the flow to perform SNAT for the cross-cluster connections sourced from local Pods,
```

Comment:
I did not get the difference from the previous flow. Is it only for connections from a remote Node?

Reply:
This will also happen for a local Pod or local Node in Gateway if user is trying to access the Endpoint of MC Service directly instead of MC Service's ClusterIP.

Reply:
Then let us make it clear: "for cross-cluster connections sourced from remote Nodes". If a local Pod on the Gateway Node trying to reach a remote multi-cluster Service Endpoint directly, the connection will hit the flow too".

Reply:
A kind reminder ^^

Reply:
Done